fix(jsx-sort-props): enforce prop sort when reservedFirst is enabled

tagoro9 opened the pull request:
If reservedFirst is enabled and noSortAlphabetically is off, then alphabetical order is only enforced in the reserved props but not in the other props. This PR fixes this.

If existing behavior is the desired, please ignore this PR.

Fixes #1320

ljharb commented on tests/lib/rules/jsx-sort-props.js:
```diff
@@ -202,9 +202,19 @@ ruleTester.run('jsx-sort-props', rule, {
       code: '<App dangerouslySetInnerHTML={{__html: "EPR"}} key={2} b />',
       options: reservedFirstAsBooleanArgs,
-      errors: [expectedReservedFirstError]
+      errors: [expectedReservedFirstError, expectedError]
```
why would this suddenly start giving two errors?

tagoro9 replied:
I think that after the changes this is triggering the `expectedError` as the `||
 (!previousIsReserved && !currentIsReserved)` check and it is true when evaluating `dangerouslySetInnerHTML` and `b`.

ljharb replied:
We should try to avoid generating the extra error.

tagoro9 commented:
I updated the tests with the expected output, but I the fix is not taking into account the reserved words:

Expected: `<App key={2} b dangerouslySetInnerHTML={{__html: "EPR"}} />`
Actual : `<App b dangerouslySetInnerHTML={{__html: "EPR"}} key={2} />`

I'm gonna try to see if I can add a fix for this.

ljharb commented on tests/lib/rules/jsx-sort-props.js:
```diff
@@ -202,9 +202,19 @@ ruleTester.run('jsx-sort-props', rule, {
+    {
+      code: '<App b a />',
+      options: reservedFirstAsBooleanArgs,
+      errors: [expectedError]
```
please add `output: '<App a b />'`

ljharb commented on tests/lib/rules/jsx-sort-props.js:
```diff
@@ -202,9 +202,19 @@ ruleTester.run('jsx-sort-props', rule, {
     {
+      code: '<App key={2} b a />',
+      options: reservedFirstAsBooleanArgs,
+      errors: [expectedError]
```
Please add `output: <App key={2} a b />` (assuming that's the expected output)

tagoro9 commented:
Sorry I took so long to take a look at the change requests.

* This [test](https://github.com/yannickcr/eslint-plugin-react/pull/1340/files#diff-85e28e7b3c88362eeaef8cf73f4ad42aR284) is no longer expecting 2 errors.
* Added expected output for the new test cases.
* With the expected output the tests were failing, so I updated the `generateFixerFunction` to take into account the `reservedFirst` config and sort props accordingly. Is this something wanted?

ljharb commented:
Definitely wanted, thanks :-)

ljharb commented on tests/lib/rules/jsx-sort-props.js:
```diff
@@ -269,8 +269,21 @@ ruleTester.run('jsx-sort-props', rule, {
+    {
       code: '<App dangerouslySetInnerHTML={{__html: "EPR"}} key={2} b />',
       options: reservedFirstAsBooleanArgs,
+      output: '<App key={2} b dangerouslySetInnerHTML={{__html: "EPR"}} />',
```
hmm - with reservedFirst tho, the expected output should be `<App dangerouslySetInnerHTML={{__html: "EPR"}} key={2} b />`, no?

tagoro9 replied:
I think `dangerouslySetInnerHTML` is not a reserved prop name for this component. I think it is only considered reserved for DOM components and not react components. This [code](https://github.com/yannickcr/eslint-plugin-react/blob/master/lib/rules/jsx-sort-props.js#L208) is deleting `dangerouslySetInnerHTML` for the cases where the component is not a DOM component.

ljharb replied:
ahhhh true! good call :-) could you add an "e" prop as well, so it's clear that it's being sorted as a regular property?

tagoro9 replied:
Added.

ljharb commented:
Planning to merge this soon.

tobyf93 commented:
Is there a reason why this change hasn't been merged in?

tobyf93 commented:
Any idea on when the merge will happen @ljharb?
